server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java

```diff
@@ -263,48 +264,52 @@ public long requireBuffer(int requireSize) {
     return requireId;
   }
 
-  public byte[] getFinishedBlockIds(
-      String appId, Integer shuffleId, Integer partitionId) throws IOException {
+  public byte[] getFinishedBlockIds(String appId, Integer shuffleId, Set<Integer> partitions) throws IOException {
     refreshAppId(appId);
-    Storage storage = storageManager.selectStorage(new ShuffleDataReadEvent(appId, shuffleId, partitionId));
-    // update shuffle's timestamp that was recently read.
-    storage.updateReadMetrics(new StorageReadMetrics(appId, shuffleId));
-
+    for (int partitionId : partitions) {
+      Storage storage = storageManager.selectStorage(new ShuffleDataReadEvent(appId, shuffleId, partitionId));
+      // update shuffle's timestamp that was recently read.
+      storage.updateReadMetrics(new StorageReadMetrics(appId, shuffleId));
+    }
     Map<Integer, Roaring64NavigableMap[]> shuffleIdToPartitions = partitionsToBlockIds.get(appId);
     if (shuffleIdToPartitions == null) {
       return null;
     }
+
     Roaring64NavigableMap[] blockIds = shuffleIdToPartitions.get(shuffleId);
     if (blockIds == null) {
       return new byte[]{};
     }
-    Roaring64NavigableMap bitmap = blockIds[partitionId % blockIds.length];
-    if (bitmap == null) {
-      return new byte[]{};
+    Map<Integer, Set<Integer>> bitmapIndexToPartitions = Maps.newHashMap();
+    for (int partitionId : partitions) {
+      int bitmapIndex = partitionId % blockIds.length;
+      if (bitmapIndexToPartitions.containsKey(bitmapIndex)) {
+        bitmapIndexToPartitions.get(bitmapIndex).add(partitionId);
+      } else {
+        HashSet<Integer> newHashSet = Sets.newHashSet(partitionId);
+        bitmapIndexToPartitions.put(bitmapIndex, newHashSet);
+      }
     }
 
-    if (partitionId > Constants.MAX_PARTITION_ID) {
-      throw new RuntimeException("Get invalid partitionId[" + partitionId
-          + "] which greater than " + Constants.MAX_PARTITION_ID);
+    Roaring64NavigableMap res = Roaring64NavigableMap.bitmapOf();
+    for (Map.Entry<Integer, Set<Integer>> entry : bitmapIndexToPartitions.entrySet()) {
+      Set<Integer> requestPartitions = entry.getValue();
+      Roaring64NavigableMap bitmap = blockIds[entry.getKey()];
+      res.or(getBlockIdsByPartitionId(requestPartitions, bitmap));
     }
-
-    return RssUtils.serializeBitMap(getBlockIdsByPartitionId(partitionId, bitmap));
+    return RssUtils.serializeBitMap(res);
   }
 
   // partitionId is passed as long to calculate minValue/maxValue
-  protected Roaring64NavigableMap getBlockIdsByPartitionId(long partitionId, Roaring64NavigableMap bitmap) {
+  protected Roaring64NavigableMap getBlockIdsByPartitionId(Set<Integer> requestPartitions,
```

Comment:
It seems not effective enough. Could we add the block id to the bitmap directly? Now we add the block id to a bitmap and then we use `or` operation to add them to the another bitmap.